File: dagger_app/src/main/java/my/liewjuntung/daggerviperusage/dagger/AppComponent.java
package my.liewjuntung.daggerviperusage.dagger;

import android.app.Application;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjectionModule;
import my.liewjuntung.daggerviperusage.app.ViperUsageApplication;
import my.liewjuntung.daggerviperusage.dagger.api.ApiModule;

/**
 * Created by pandawarrior on 24/10/2017.
 */
@Singleton
@Component(modules = {AndroidInjectionModule.class, AppModule.class, ActivityBuilder.class, ApiModule.class})
public interface AppComponent {

    @Component.Builder interface Builder {
        AppComponent build();
        @BindsInstance
        Builder application(Application application);

    }
    /**
     * Create an interface for {@link my.liewjuntung.daggerviperusage.app.ViperUsageApplication} to be injected
     */
    void inject(ViperUsageApplication application);

}

File: kotlin_dagger_app/src/main/java/my/liewjuntung/kotlindaggerapp/dagger/AppComponent.kt
package my.liewjuntung.daggerviperusage.dagger

import android.app.Application
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjectionModule
import my.liewjuntung.kotlindaggerapp.app.ViperUsageApplication
import my.liewjuntung.kotlindaggerapp.dagger.ActivityBuilder
import my.liewjuntung.kotlindaggerapp.dagger.AppModule
import my.liewjuntung.kotlindaggerapp.dagger.api.ApiModule
import javax.inject.Singleton

/**
 * Created by pandawarrior on 24/10/2017.
 */
@Singleton
@Component(modules = arrayOf(AndroidInjectionModule::class, AppModule::class, ActivityBuilder::class, ApiModule::class))
interface AppComponent {

    @Component.Builder
    interface Builder {
        fun build(): AppComponent
        @BindsInstance
        fun application(application: Application): Builder
    }

    /**
     * Create an interface for [my.liewjuntung.kotlindaggerapp.app.ViperUsageApplication] to be injected
     */
    fun inject(application: ViperUsageApplication)

}
